### ISSUE TITLE
Rails 6

### ISSUE THREAD

**Issue by aiomaster** [refinery/refinerycms#3439]
Hi,

do you see some show stoppers regarding rails 6 upgrade?
Is there interest?
Was some work already done somewhere?

Thanks!

**Comment by angarc** [comment]
Also curious about this.

**Comment by joebutler2** [comment]
I'm attempting to use it on a greenfield Rails 6 app, but there is a dependency conflict with ActionPack.

```Bundler could not find compatible versions for gem "actionpack":
  In Gemfile:
    rails (~> 6.0) was resolved to 6.0.0, which depends on
      actionpack (= 6.0.0)

    refinerycms was resolved to 4.0.2, which depends on
      refinerycms-core (= 4.0.2) was resolved to 4.0.2, which depends on
        actionpack (>= 5.2.0, < 6)```

Hopefully this is a trivial fix and that we can update refinerycms-core to work with ActionPack 6. 

I'll update this issue as I learn more.

**Comment by joebutler2** [comment]
Here's the branch I created: https://github.com/joebutler2/refinerycms/pull/new/rails-6-support

Currently there are 15 failing tests. I'll look into it tomorrow. https://gist.github.com/joebutler2/54c455633ae5fb9288fc0ec8e0a25c00

**Comment by joebutler2** [comment]
I'm requesting help on this now. Progress has been made but there are a number of gems we had to upgrade in tandem.

Right now, I'm working through a failing test that's related to how image uploads work in Selenium. The feature works fine when manually testing it in the generated dummy app.

Here is a screenshot of the error occurring in Selenium (a la `save_and_open_screenshot`).

![capybara-201909201605507514458208](https://user-images.githubusercontent.com/6955350/65357258-88667600-dbc4-11e9-9a7b-37eae7ec3e2f.png)

Here is the stacktrace from test failure (note the line numbers will be off since I adjusted the formatting and added debug code locally)

```
Shared Example Group: "uploads images" called from ./images/spec/features/refinery/admin/images_spec.rb:16
     # ./images/spec/support/shared_examples/image_uploader.rb:29:in `block (3 levels) in <top (required)>'
     # ./core/spec/support/database_cleaner.rb:18:in `block (2 levels) in <top (required)>'
```


When I tested it manually I used the same image that is used in the automated test. `spec/fixtures/image-with-dashes.jpg` 

Here's the branch, if anyone has the time to help out (or pair program!) it would be appreciated.

https://github.com/joebutler2/refinerycms/commits/rails-6-support

**Comment by anitagraham** [comment]
Quick suggestion: it’s failing on the test for a valid mime type. 

 

I would (temporarily) add to the test something to tell you what the image mime type is recognized as.

 

For the sake of completeness, if nothing else, I would also carry through the change from “whitelisted” to “allowed” to the validation. 

 

I will try to have a closer look/test but I am about to go away for three weeks, so might not get the time.

 

-- 

Anita Graham

anita@joli.com.au

0411 645 149

**Comment by anitagraham** [comment]
@joebutler2 
I downloaded your branch, did a `bundle update` and ran the images tests which all passed.
This was still on Rails 5.2.3. Trying to go to Rails 6 triggered lots of dependency issues. 

In the meantime I added a line to the test which failed for you: 
```
#images/spec/support/shared_examples/image_uploader.rb
      fill_in  'image_image_title', with: 'Image With Dashes'
      fill_in  'image_image_alt', with: "Alt description for image"
      puts "----- Image path is #{image_path}"
      click_button ::I18n.t('save', scope: 'refinery.admin.form_actions')
    end
```
I guess you could check that `::Refinery::Images.whitelisted_mime_types` has the expected mime types, as that is the other half of the validation.

**Comment by anitagraham** [comment]
I've submitted a PR (#3440) to refinery/refinerycms which incidentally gives a more specific message when an image with an invalid mime-type is uploaded. 

This might be helpful finding the source of the test failure.

btw: what was the actual error. There isn't enough stacktrace there to see the error message.

**Comment by joebutler2** [comment]
@anitagraham thank you for helping out. It turns out most of those errors were due to my local environment. This is now an open pull request #3441. As of right now there is only one failure in CI, and that's in the **pages** extension when running **mysql** in **ruby 2.4.2** ([link to failed build](https://travis-ci.com/refinery/refinerycms/jobs/238020422)). The other errors in the build are the same ones we are currently seeing in master.

Here is the stack trace for that failure:

```
  1) Pages new/create allows to easily create nested page
1405     Failure/Error: find("a[href='#{refinery.new_admin_page_path(:parent_id => parent_page.id)}']").click
1406     
1407     Capybara::ElementNotFound:
1408       Unable to find css "a[href='/refinery/pages/new?parent_id=15']"
1409     # ./pages/spec/features/refinery/admin/pages_spec.rb:199:in `block (3 levels) in <module:Admin>'
1410     # -e:1:in `<main>'
```

---

As for handling the dependencies, the branch featured in that PR should contain most of the dependency updates.

My local environment mirrors what heroku currently supports (some specific to me). Ruby 2.5.5 and postgresql 11.5, and Rails 6.0.

We also need to update the `.gemspec` file in `refinerycms-i18n` since that is blocking us from Rails 6 as well. Here is an [open PR for that repo](https://github.com/refinery/refinerycms-i18n/pull/80). I'll take a look into why that build failed tomorrow.

Once we fix both of these failed builds, refinery should be good to go for Rails 6 apps. Of course, I'll report and help fix any bugs that arise.

If anyone wants to help with either one of those pull requests, I would love to collaborate and possibly do remote pair programming (via [Tuple](http://tuple.app) or [Floobits](http://floobits.com)). /cc @aiomaster @angarc

**Comment by joebutler2** [comment]
I re-triggered the build for PR #3441 and that failing test passed. So from my point of view, that branch should be good to merge into master, although my preference is to fix master ASAP.

I'm dedicating some time right now to look into the failure in master.

Also, I still need to figure out why the build failed for updating `refinerycms-i18n` to support Rails 6 ([pull request link](https://github.com/refinery/refinerycms-i18n/pull/80)). 

Hopefully, we can resolve these broken builds in the next couple of days, then refinery will support Rails 6!

**Comment by joebutler2** [comment]
Quick update:

The updates to `refinerycms-i18n` for Rails 6 support are ready to be merged in. https://github.com/refinery/refinerycms-i18n/pull/80

If we could get someone to merge that in and release a patch version of the gem (@bricesanchez  @aiomaster), then we can wrap up the pull request for `refinerycms` itself https://github.com/refinery/refinerycms/pull/3441. 

I'm also hopeful that updating the `mobility` gem in `refinerycms-i18n` will also fix the build in master. See this [comment for details](https://github.com/refinery/refinerycms/pull/3441#issuecomment-535089241).

Thanks everyone who has helped out! Hopefully we can wrap this up in the coming days!

**Comment by joebutler2** [comment]
I compiled the updated version of `refinerycms-i18n` locally and now I'm receiving dependency conflicts in bundler. https://gist.github.com/joebutler2/d78c3ed245652663032c7ede84075ef8

I'm hoping to get them resolved quickly!

**Comment by joebutler2** [comment]
I was able to resolve the `refinerycms-i18n` conflict locally (thanks to `geminabox`!). 

Now once https://github.com/refinery/refinerycms-i18n/pull/80 is merged and the patch version of the gem is released we will be very close to having RefineryCMS support Rails 6!

Also, I updated this gist to the remaining dependency conflicts, hopefully, we can wrap it up tomorrow! 
https://gist.github.com/joebutler2/d78c3ed245652663032c7ede84075ef8

**Comment by joebutler2** [comment]
Was able to resolve all of the dependency conflicts locally (with a custom rubygem server). Now there is a loading issue with `refinerycms-core`.

The PR for `refinerycms-i18n` was accepted and merged in by Philip (thanks parndt!) Waiting for that to be released. Also had to update the `decorators` gem as well (See this [PR for details](https://github.com/parndt/decorators/pull/19)).

I'll keep working on the pull request while figuring the rest out. I'll stop updating this issue going forward, if you want to see notifications / progress watch #3441.

**Comment by parndt** [comment]
Also see #3446

**Comment by parndt** [comment]
Fixed by #3446
